cql3/expr: remove expr::token

cvybhu opened the pull request:
Let's remove `expr::token` and replace all of its functionality with `expr::function_call`.

`expr::token` is a struct whose job is to represent a partition key token.
The idea is that when the user types in `token(p1, p2) < 1234`, this will be internally represented as an expression which uses `expr::token` to represent the `token(p1, p2)` part.

The situation with `expr::token` is a bit complicated.
On one hand side it's supposed to represent the partition token, but sometimes it's also assumed that it can represent a generic call to the `token()` function, for example `token(1, 2, 3)` could be a `function_call`, but it could also be `expr::token`.

The query planning code assumes that each occurence of expr::token
represents the partition token without checking the arguments.
Because of this allowing `token(1, 2, 3)` to be represented as `expr::token` is dangerous - the query planning might think that it is `token(p1, p2, p3)` and plan the query based on this, which would be wrong.

Currently `expr::token` is created only in one specific case.
When the parser detects that the user typed in a restriction which has a call to `token` on the LHS it generates `expr::token`.
In all other cases it generates an `expr::function_call`.
Even when the `function_call` represents a valid partition token, it stays a `function_call`. During preparation there is no check to see if a `function_call` to `token` could be turned into `expr::token`. This is a bit inconsistent - sometimes `token(p1, p2, p3)` is represented as `expr::token` and the query planner handles that, but sometimes it might be represented as `function_call`, which the query planner doesn't handle.

There is also a problem because there's a lot of code duplication between a `function_call` and `expr::token`.
All of the evaluation and preparation is the same for `expr::token` as it's for a `function_call` to the token function.
Currently it's impossible to evaluate `expr::token` and preparation has some flaws, but implementing it would basically consist of copy-pasting the corresponding code from token `function_call`.

One more aspect is multi-table queries.
With `expr::token` we turn a call to the `token()` function into a struct that is schema-specific.
What happens when a single expression is used to make queries to multiple tables? The schema is different, so something that is represented as `expr::token` for one schema would be represented as `function_call` in the context of a different schema.
Translating expressions to different tables would require careful manipulation to convert `expr::token` to `function_call` and vice versa. This could cause trouble for index queries.

Overall I think it would be best to remove `expr::token`.

Although having a clear marker for the partition token is sometimes nice for query planning, in my opinion the pros are outweighted by the cons.
I'm a big fan of having a single way to represent things, having two separate representations of the same thing without clear boundaries between them causes trouble.

Instead of having both `expr::token` and `function_call` we can just have the `function_call` and check if it represents a partition token when needed.

Refs: #12906
Refs: #12677
Closes: #12905

cvybhu commented:
I explored keeping `expr::token` in #12677 and also tried a hybrid approach by having `struct token {function_call tok_call;}` on the branch https://github.com/cvybhu/scylladb/tree/cql-token-effort.
After trying all of these options I'm leaning towards just removing it altogether.

nyh commented:
I'll start reviewing this series now. You mentioned "Fixes https://github.com/scylladb/scylladb/issues/12906" - did you really mean it is fully fixed by this series?

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -908,6 +908,17 @@ test_assignment_function_call(const cql3::expr::function_call& fc, data_dictiona
+    if (receiver.type->underlying_type() == expr_type->underlying_type()) {
+        return assignment_testable::test_result::EXACT_MATCH;
+    } else if (receiver.type->is_value_compatible_with(*expr_type)) {
+        return assignment_testable::test_result::WEAKLY_ASSIGNABLE;
```
I think this is correct and consistent with the existing code, but wrong in general. We shouldn't allow assignment from long to timestamp type without a cast. However it will be difficult to deprecate such assignments.

cvybhu replied:
AFAIR in another discussion we decided to allow them for now, for compatibility with Cassandra, but eventually we will add a flag that forbids such casts, something like `strict_types: true`.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -817,17 +817,36 @@ cast_prepare_expression(const cast& c, data_dictionary::database db, const sstri
+        // The information about keyspace is provided as an argument, but the information about column_family
+        // has to be acquired either from schema_opt or receiver.
+        // When both of them are not available it's impossible to prepare the function.
         return std::nullopt;
```
That's only true for the token function, no?

cvybhu replied:
Right, I will change it so that the column family is required only for the `token()` function.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -817,17 +817,36 @@ cast_prepare_expression(const cast& c, data_dictionary::database db, const sstri
+        if (prepared_arg_opt.has_value()) {
+            partially_prepared_args.emplace_back(*prepared_arg_opt);
+        } else {
+            partially_prepared_args.push_back(argument);
```
I feel that it's wrong to lose information about whether an argument was prepared or not. Won't we prepare it again?

cvybhu replied:
I added this because I ran into problems with `unresolved_identifier` and `constant`. `functions::get` requires a list of function arguments, where each argument is `assignment_testable`. `unresolved_identifier` can't be `assignment_testable`, because it represents an unknown column, there's just the column name.

To fix it I prepare the arguments that can be prepared without a receiver, which includes `unresolved_identifiers`. I could potentially prepare only the occurrences of `unresolved_identifier`, but this felt like a more general solution.

We will prepare some of them again, but is it wrong to do so?
I took a look at `prepare_expression` and it looks like it's able to reprepare a previously prepared expression for all the cases that matter.

I think that `prepare_expression` should be robust enough that calling it twice isn't a bug.  It's hard to keep track of which expressions were already prepared and which ones were not, I bet we already call prepare twice somewhere by accident.

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -855,11 +874,11 @@ prepare_function_call(const expr::function_call& fc, data_dictionary::database d
-                                                functions::functions::make_arg_spec(receiver->ks_name, receiver->cf_name, *scalar_fun, i));
+    for (size_t i = 0; i < partially_prepared_args.size(); ++i) {
+        expr::expression e = prepare_expression(partially_prepared_args[i], db, keyspace, schema_opt,
+                                                functions::functions::make_arg_spec(keyspace, cf_name, *scalar_fun, i));
```
Yes, here we do so.

avikivity commented on cql3/expr/expression.cc:
```diff
@@ -2694,5 +2694,26 @@ adjust_for_collection_as_maps(const expression& e) {
+                                      [](const shared_ptr<functions::function>& fun) { return fun->name(); }},
+                   fun_call.func);
+
+    return fun_name.has_keyspace() ? fun_name == token_function_name : fun_name.name == token_function_name.name;
```
What if someone defines my_keyspace.token? Is it legal?

avikivity replied:
Also, can we every see my_keyspace.token? Won't token(x) be converted to system.token(x) during parsing?

avikivity commented:
Looks good.

nyh commented on cql3/expr/expression.cc:
```diff
@@ -1123,14 +1123,18 @@ std::ostream& operator<<(std::ostream& os, const expression::printer& pr) {
-                        fmt::print(os, "<anonymous function>({})", fmt::join(fc.args | transformed(to_printer), ", "));
-                    },
-                }, fc.func);
+                if (is_token_function(fc)) {
```
I think this fix Refs https://github.com/scylladb/scylladb/issues/10632#issuecomment-1374875396, or maybe if you fix one more small thing, it can Fixes it.

nyh commented on cql3/expr/expression.cc:
```diff
@@ -1123,14 +1123,18 @@ std::ostream& operator<<(std::ostream& os, const expression::printer& pr) {
+                            fmt::print(os, "{}({})", named, fmt::join(fc.args | transformed(to_printer), ", "));
+                        },
+                        [&] (const shared_ptr<functions::function>& anon) {
+                            fmt::print(os, "<anonymous function>({})", fmt::join(fc.args | transformed(to_printer), ", "));
```
I'm not a fan of this "anonymous function" (see also https://github.com/scylladb/scylladb/issues/10632#issuecomment-1374875396). After all, the user somehow called this function in CQL, so there must be some CQL string that can be remembered together with the call and be used when we want to print this call.

But I guess this is a pre-existing problem. At least you are fixing it for token() which is what bothered me in that issue. I don't know in what other cases it can appear.

cvybhu replied:
I'm not sure why the original code had the `<anonymous function>`. functions::function has a name method: 
```c++
virtual const function_name& name() const = 0;
```

Maybe some internal functions have empty names because they were never meant to be printed? Then the user would get confused because they see `(arg)` instead of `<anonymous function>(arg)`. This is something that would be good to fix, but probably not in this PR.

nyh commented on cql3/expr/expression.cc:
```diff
@@ -793,7 +775,11 @@ static value_set possible_lhs_values(const column_definition* cdef,
                             return unbounded_value_set;
                         },
-                        [&] (token) -> value_set {
+                        [&] (const function_call& token_fun_call) -> value_set {
```
nitpick: at this point you don't know that token_fun_call is a token function call (you'll check this next), so maybe better to call this variable just fun_call?

cvybhu replied:
I wanted to indicate that the only acceptable `function_call` here is a call to the `token()` function. Any other `function_call` that happens to be there means that there's a bug and it will trigger an `on_internall_error`.
In that sense I know that this is a call to the token() function - anything else would mean that there's a bug.

nyh commented on cql3/expr/expression.cc:
```diff
@@ -793,7 +775,11 @@ static value_set possible_lhs_values(const column_definition* cdef,
-                        [&] (token) -> value_set {
+                        [&] (const function_call& token_fun_call) -> value_set {
+                            if (!is_partition_token_for_schema(token_fun_call, *table_schema_opt)) {
+                                on_internal_error(expr_logger, "possible_lhs_values: function calls are not supported as the LHS of a binary expression");
```
Maybe add the words "... except token() ..." to this message, since this sort of function?
I also had another thought - maybe it will be nicer to check three cases:
1. If token(p1,p2,p3), it's supported
2. If token(something,else), print an error that only token of full partition key is allowed in LHS
3. If someotherfunction(), print an error that someotherfunction() is not allowed in LHS, the only function allowed is token().

cvybhu replied:
It's an `on_internal_error` message, so it is never really meant to be printed. I don't know if there's any reason to make it pretty - once it triggers someone will go and have to do some debugging anyway, and the message is unique, so it's obvious where the error comes from.

nyh reviewed (approved):
Very nice series. Took me a long time to read, but everything was explained very well. I left some comments and requests, but nothing is really blocking.

cvybhu commented:
> What if someone defines my_keyspace.token? Is it legal?

> Can a SELECT on table X.Y use Z.token(...) (where Z is a different keyspace)? How does it work? If not, why is "system.token(...)" valid?

I tested whether defining an UDF with the name `token` works, and it looks like it's behaving reasonably.
It's possible to create the UDF and refer to it as `ks.token`. It's also possible to use `ks2.token` for a query to a table that is inside `ks`:


```cql
cqlsh> CREATE KEYSPACE IF NOT EXISTS ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};
cqlsh> CREATE KEYSPACE IF NOT EXISTS ks2 WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};
cqlsh> CREATE TABLE ks.t (p int primary key);
cqlsh> CREATE FUNCTION ks.token(pk int) RETURNS NULL ON NULL INPUT RETURNS bigint LANGUAGE LUA AS 'return pk + 1;';
cqlsh> CREATE FUNCTION ks2.token(pk int) RETURNS NULL ON NULL INPUT RETURNS bigint LANGUAGE LUA AS 'return pk + 2;';
cqlsh> INSERT INTO ks.t (p) VALUES (1);
cqlsh> SELECT p, token(p) FROM ks.t;

 p | system.token(p)
---+----------------------
 1 | -4069959284402364209

(1 rows)
cqlsh> SELECT p, system.token(p) FROM ks.t;

 p | system.token(p)
---+----------------------
 1 | -4069959284402364209

(1 rows)
cqlsh> SELECT p, ks.token(p) FROM ks.t;

 p | ks.token(p)
---+-------------
 1 |           2

(1 rows)
cqlsh> SELECT p, ks2.token(p) FROM ks.t;

 p | ks2.token(p)
---+--------------
 1 |            3

(1 rows)
```

cvybhu commented on cql3/expr/expression.cc:
```diff
@@ -2694,5 +2694,26 @@ adjust_for_collection_as_maps(const expression& e) {
+                                      [](const shared_ptr<functions::function>& fun) { return fun->name(); }},
+                   fun_call.func);
+
+    return fun_name.has_keyspace() ? fun_name == token_function_name : fun_name.name == token_function_name.name;
```
The name `token` gets converted to `system.token` during parsing. The parser has a special rule just for the token() function call:
```c++
    | K_TOKEN l=tupleOfIdentifiers type=relationType t=term
        {
          $e = binary_operator(
            function_call{functions::function_name::native_function("token"), std::move(l.elements)},
            type,
            std::move(t));
        }
```

`native_function("token")` corresponds to the name `system.token`.

I'm not sure if it's possible to end up with a function name that doesn't have a keyspace, but I guess it's better to handle it just in case. I took this code from `functions::get`.

cvybhu commented:
> You mentioned "Fixes #12906" - did you really mean it is fully fixed by this series?

At the moment of making this PR I thought it did fully fix #12906, after this change the only expressions that can't be evalutated would be:
```c++
        [](const unresolved_identifier&) -> cql3::raw_value {
            on_internal_error(expr_logger, "Can't evaluate unresolved_identifier");
        },
        [](const column_mutation_attribute&) -> cql3::raw_value {
            on_internal_error(expr_logger, "Can't evaluate a column_mutation_attribute");
        },
        [](const field_selection&) -> cql3::raw_value {
            on_internal_error(expr_logger, "Can't evaluate a field_selection");
        },
        [](const untyped_constant&) -> cql3::raw_value {
            on_internal_error(expr_logger, "Can't evaluate a untyped_constant ");
        },
```

The fact that we can't evaluate `unresolved_identifier`, `column_mutation_attribute` and `untyped_constant` makes sense - they don't occur in prepared expressions. However now I have doubts about `field_selection`. At first glance I thought that this is something from the selectables that I don't have to worry about, but after a closer look it looks like this is used to select a field of an UDT value. This is something that should eventually be evaluable, so that we can have restrictions like `my_udt_col.field1 = 123`.
OTOH we currently such expressions can't occur in the WHERE clause, the parser doesn't handle that, so it's not really a problem.
But #12906 is about being able to evaluate ALL expressions that should be evaluable, so I will remove the `Fixes:`.

cvybhu commented on cql3/expr/prepare_expr.cc:
```diff
@@ -817,17 +817,36 @@ cast_prepare_expression(const cast& c, data_dictionary::database db, const sstri
+        // which define the environment where this function is defined and executed.
+        // The information about keyspace is provided as an argument, but the information about column_family
+        // has to be acquired either from schema_opt or receiver.
+        // When both of them are not available it's impossible to prepare the function.
```
Right, I will change it so that the column family is required only for the `token()` function.

cvybhu commented on cql3/expr/expression.hh:
```diff
@@ -642,7 +642,7 @@ inline bool is_multi_column(const binary_operator& op) {
 }
 
-inline bool has_token(const expression& e) {
+inline bool has_partition_token(const expression& e, const schema& table_schema) {
```
Added a comment.

I agree that the name `has_` is a bit unfortunate, it suggests that we do a search over all the expression AST nodes.

I think it originally came from a statement_restrictions method that was named in the same way.

My hope is that it will later disappear when I get to rewriting `statement_restrictions`.
Right now it's used to see if the restrictions are `partition_token = ...`, but it's a bit fragile, I think query analysis should generally be changed to rely more on pattern matching, and less on random functions and allowing only a strict subset of restrictions to be written.

cvybhu commented:
v2:
* rebased onto the newest master, conflicting changes were introduced since the original version
* changed argument name of `is_token_function` from `fun_expr` to `e`
* added a comment for `has_partition_token`
* stopped requiring column family to prepare every `function_call`, it's now only required for `token()`
* modified `get_lhs_receiver` and `validate_and_prepare_new_restrictions` to also handle `token(pk, pk)`, because doing so caused `on_internal_error`
* modified a boost test, it complained that there's no `operator<<` for  `function_name`
* modified a recent test by Kefu Chai, the error messages changed a bit because `token()` is now handled like the other functions

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/971/

nyh commented:
> The fact that we can't evaluate `unresolved_identifier`, `column_mutation_attribute` and `untyped_constant` makes sense - they don't occur in prepared expressions. However now I have doubts about `field_selection`. At first glance I thought that this is something from the selectables that I don't have to worry about, but after a closer look it looks like this is used to select a field of an UDT value. This is something that should eventually be evaluable, so that we can have restrictions like `my_udt_col.field1 = 123`.

This is indeed still a bug, I opened a bug on it just last week: https://github.com/scylladb/scylladb/issues/13624
